### ISSUE TITLE
docker images for arm32 fail to run maven builds since JDK 17

### ISSUE THREAD

**Issue by easimon** [adoptium/temurin-build#3094]
**What are you trying to do?**

Trying to build a multi-arch docker images based on eclipse-temurin:17 for `linux/amd64`, `linux/arm64/v8`, `linux/arm/v7`

**Expected behaviour:**

The build succeeds.

**Observed behaviour:**

For `eclipse-temurin:8` and `eclipse-temurin:11`, everything works as expected for all platforms, 
`eclipse-temurin:17` and `eclipse-temurin:18` fail to build (`./mvnw package`) on `linux/arm/v7` with a stack trace like the following: 

```
[builder 6/6] RUN ./mvnw -B package
2.971 Exception in thread "main" java.lang.ExceptionInInitializerError
2.981 	at java.base/javax.crypto.Cipher.getInstance(Cipher.java:546)
2.982 	at java.base/sun.security.ssl.SSLCipher.isTransformationAvailable(SSLCipher.java:511)
2.982 	at java.base/sun.security.ssl.SSLCipher.<init>(SSLCipher.java:500)
2.983 	at java.base/sun.security.ssl.SSLCipher.<clinit>(SSLCipher.java:82)
2.983 	at java.base/sun.security.ssl.CipherSuite.<clinit>(CipherSuite.java:65)
2.984 	at java.base/sun.security.ssl.SSLContextImpl.getApplicableSupportedCipherSuites(SSLContextImpl.java:343)
2.985 	at java.base/sun.security.ssl.SSLContextImpl$AbstractTLSContext.<clinit>(SSLContextImpl.java:556)
2.985 	at java.base/java.lang.Class.forName0(Native Method)
2.986 	at java.base/java.lang.Class.forName(Class.java:375)
2.986 	at java.base/java.security.Provider$Service.getImplClass(Provider.java:1933)
2.987 	at java.base/java.security.Provider$Service.getDefaultConstructor(Provider.java:1964)
2.987 	at java.base/java.security.Provider$Service.newInstanceOf(Provider.java:1878)
2.988 	at java.base/java.security.Provider$Service.newInstanceUtil(Provider.java:1886)
2.988 	at java.base/java.security.Provider$Service.newInstance(Provider.java:1861)
2.989 	at java.base/sun.security.jca.GetInstance.getInstance(GetInstance.java:236)
2.989 	at java.base/sun.security.jca.GetInstance.getInstance(GetInstance.java:164)
2.990 	at java.base/javax.net.ssl.SSLContext.getInstance(SSLContext.java:185)
2.990 	at java.base/javax.net.ssl.SSLContext.getDefault(SSLContext.java:110)
2.991 	at java.base/javax.net.ssl.SSLSocketFactory.getDefault(SSLSocketFactory.java:83)
2.991 	at java.base/javax.net.ssl.HttpsURLConnection.getDefaultSSLSocketFactory(HttpsURLConnection.java:336)
2.992 	at java.base/javax.net.ssl.HttpsURLConnection.<init>(HttpsURLConnection.java:292)
2.992 	at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.<init>(HttpsURLConnectionImpl.java:81)
2.993 	at java.base/sun.net.www.protocol.https.Handler.openConnection(Handler.java:62)
2.993 	at java.base/sun.net.www.protocol.https.Handler.openConnection(Handler.java:57)
2.994 	at java.base/java.net.URL.openConnection(URL.java:1094)
2.995 	at org.apache.maven.wrapper.DefaultDownloader.downloadInternal(DefaultDownloader.java:104)
2.995 	at org.apache.maven.wrapper.DefaultDownloader.download(DefaultDownloader.java:97)
2.996 	at org.apache.maven.wrapper.Installer.createDist(Installer.java:94)
2.996 	at org.apache.maven.wrapper.WrapperExecutor.execute(WrapperExecutor.java:151)
2.997 	at org.apache.maven.wrapper.MavenWrapperMain.main(MavenWrapperMain.java:76)
2.998 Caused by: java.lang.SecurityException: Can not initialize cryptographic mechanism
2.999 	at java.base/javax.crypto.JceSecurity.<clinit>(JceSecurity.java:119)
2.999 	... 30 more
3.001 Caused by: java.lang.SecurityException: Couldn't iterate through the jurisdiction policy files: unlimited
3.001 	at java.base/javax.crypto.JceSecurity.setupJurisdictionPolicies(JceSecurity.java:372)
3.002 	at java.base/javax.crypto.JceSecurity$1.run(JceSecurity.java:110)
3.002 	at java.base/javax.crypto.JceSecurity$1.run(JceSecurity.java:107)
3.003 	at java.base/java.security.AccessController.doPrivileged(AccessController.java:569)
3.003 	at java.base/javax.crypto.JceSecurity.<clinit>(JceSecurity.java:106)
3.004 	... 30 more
```

**Any other comments:**

I have a minimal bug reproducer at https://github.com/easimon/bug-reproducers/tree/eclipse-temurin-arm32, and a pipeline run that shows the observed behaviour at https://github.com/easimon/bug-reproducers/actions/runs/3013711874.

When building the JAR with a working version (e.g. 11), I can package it into a working final image based on 17 or 18-jre. Just building fails.

<img width="323" alt="image" src="https://user-images.githubusercontent.com/1749657/189082761-46cdb4b9-3271-40c1-86ca-44468393d0a9.png">

**Comment by karianna** [comment]
@easimon Can you confirm that those images contain version 17.0.4.1?

**Comment by easimon** [comment]
I do not have a arm32 machine at hand to test it, the builds are using `docker buildx build --platform` for cross-platform builds. But both on Github Actions (some Intel Xeon, amd64) and locally (M1 Mac, arm64), the builds fail the same way.

I did not specify a minor/patch version, but just `eclipse-temurin:$MAJOR`, in 8, 11, 17, 18.

https://github.com/easimon/bug-reproducers/actions/runs/3016801184 now logs the exact version in use, being

```
OpenJDK Server VM Temurin-17.0.4.1+1 (build 17.0.4.1+1, mixed mode, sharing)
OpenJDK Server VM Temurin-18.0.2.1+1 (build 18.0.2.1+1, mixed mode, sharing)
```

**Comment by sxa** [comment]
I'm not sure I can easily provide assistance with this. The docker images appear valid and work in a non-buildx environment. I don't have any machines with buildx installed and unfortunately I'm about to head on vacation for a week. Someone else might be able to give advice, but hopefully that might point you in the direction of something that might indicate whether buildx is at fault.
On your M1 mac you could try and run it without buildx by putting these lines at the top of your example:

```
ARG BUILD_IMAGE=eclipse-temurin:17@sha256:0a3bf35d831d8887257a89acd50243642b63a083970876bc3f11090b8efdb000
ARG RUNTIME_IMAGE=eclipse-temurin:17-jre@sha256:626ea369e3e754a89ca3724e165b3221f684d971b7439f9d839e2c67228c5e7c
```
(Those SHAs came from `docker manifest inspect eclipse-temurin:17` and `17-jre` so those lines would fix to the current version at the time of writing and not get updated)

**Comment by easimon** [comment]
Hm -- Docker Desktop for Mac comes with QEMU preconfigured for multiplatform builds, and I don't know how to disable that and force Docker to try and run an image for a different platform *without* QEMU -- if that's possible at all. I am also not sure if it would work to run arm32 binaries unemulated on M1 -- at least for Mac Apps, there is no 32bit support in the OS.

Switching to SHAs for the e.g. `linux/amd64` platform gives me a warning, 

```
[Warning] The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
```

but succeeds to build the image for that platform -- so QEMU is obviously still active no matter what I try. So if that's a buildx or qemu issue instead of an issue with the image, I can't verify it using Docker Desktop for Mac.
Trying the docker build with the arm32 SHAs from above yields the same error message on my M1.

**Comment by easimon** [comment]
Digging deeper: The stack trace looks like there's something wrong with java.crypto not being able to initialize properly, triggered by maven/mvnw that wants to download jars from Maven Central. To rule out `mvn` or `mvnw` as the culprit, I extended my test program in `Main.java` to do a HTTP(s) download as well.

Then I changed the Github Actions workflow to always build on `eclipse-temurin:11` (because that works on all archs), and package the result into a matrix of runtime images (11,17,18 plus their jre-variants). I removed JDK 8 compatibility so I can use the fluent Java 11 HTTP Client for the GET request.

https://github.com/easimon/bug-reproducers/tree/eclipse-temurin-arm32 is updated with these changes.

https://github.com/easimon/bug-reproducers/actions/runs/3021951990 is a run of that workflow. 

Read each build job like 
```
Build linux/amd64: 11/17        (Build with eclipse-temurin:11, run with eclipse-temurin:17, on linux/amd64)
Build linux/arm64/v8: 11/17-jre (Build with eclipse-temurin:11, run with eclipse-temurin:17-jre, on linux/arm64/v8)
```

The jobs that execute on `linux/arm/v7` fail to run the resulting docker image, with the stack trace from the top, EXCEPT the ones using JDK/JRE 11.

**Comment by easimon** [comment]
While I would like to provide more details, I am not sure how to come closer to the root cause, any hints welcome.

**Comment by easimon** [comment]
Update: I dug down to the line in the JDK that triggers the error (and then shadows the underlying native IO Error), and updated the example code with a stripped-down copy of the function to just display the exception.

In the errored jobs of https://github.com/easimon/bug-reproducers/actions/runs/3023586142, it shows now 

```
java.nio.file.DirectoryIteratorException: java.nio.file.FileSystemException: /opt/java/openjdk/conf/security/policy/unlimited: Value too large for defined data type
```

With the plain exception message, I was able to find similar issues in other projects, pointers to QEMU being the culprit, and also found some workarounds, described in https://github.com/easimon/bug-reproducers/tree/eclipse-temurin-arm32#readme

Anyway, to me it seems, that this is not directly an error in the JDK or the eclipse-temurin docker image, but instead either Docker, GLIBC or QEMU, or a combination of it. I did not yet find the reason why that does not happen with JDK 11 or 8 -- maybe the way JceSecurity traverses the policy directory changed, or the GLIBC version in `eclipse-temurin:11`'s base image is different -- I did not dig deeper (yet).

Anyway, I guess this issue can be closed, since the image itself is fine. It's the surrounding runtime that fails in cross-platform scenarios.
